src/notation/view/internal/dynamicpopupmodel.cpp

```diff
@@ -0,0 +1,225 @@
+        { DynamicType::OTHER,  62, 0.0, DynamicPopupModel::Decrescendo, "" },
+    },
+};
+
+DynamicPopupModel::DynamicPopupModel(QObject* parent)
+    : AbstractElementPopupModel(PopupModelType::TYPE_DYNAMIC, parent)
+{
+}
+
+QString DynamicPopupModel::fontFamily() const
+{
+    return QString::fromStdString(m_item->score()->engravingFont()->family());
+}
+
+QVariantList DynamicPopupModel::pages() const
+{
+    return m_pages;
+}
+
+QString DynamicPopupModel::xmlTextToQString(const std::string text, IEngravingFontPtr engravingFont) const
+{
+    const std::string startTag = "<sym>";
+    const std::string endTag = "</sym>";
+
+    String str;
+    size_t pos = 0;
+
+    while (true) {
+        size_t startPos = text.find(startTag, pos);
+        if (startPos == std::string::npos) {
+            break;
+        }
+        size_t symNameStart = startPos + startTag.length();
+        size_t endPos = text.find(endTag, symNameStart);
+        if (endPos == std::string::npos) {
+            break;
+        }
+        SymId id = SymNames::symIdByName(text.substr(symNameStart, endPos - symNameStart));
+        str += engravingFont->toString(id);
+        pos = endPos + endTag.length();
+    }
+
+    return str.toQString();
+}
+
+void DynamicPopupModel::init()
+{
+    AbstractElementPopupModel::init();
+
+    m_pages.clear();
+
+    if (!m_item) {
+        return;
+    }
+
+    IEngravingFontPtr engravingFont = m_item->score()->engravingFont();
+
+    for (const QList<DynamicPopupModel::PageItem>& page : DYN_POPUP_PAGES) {
+        QVariantList variantPage;
+        for (const DynamicPopupModel::PageItem& item : page) {
+            QVariantMap variantMap {
+                { "text", xmlTextToQString(item.text, engravingFont) },
+                { "width", item.width },
+                { "offset", item.offset },
+                { "type", item.itemType },
+            };
+            variantPage.append(variantMap);
+        }
+        m_pages.append(QVariant::fromValue(variantPage));
+    }
+
+    emit pagesChanged();
+}
+
+void DynamicPopupModel::addOrChangeDynamic(int page, int index)
+{
+    IF_ASSERT_FAILED(m_item) {
+        return;
+    }
+
+    if (!m_item->isDynamic()) {
+        return;
+    }
+
+    beginCommand();
+    m_item->undoChangeProperty(Pid::TEXT, Dynamic::dynamicText(DYN_POPUP_PAGES[page][index].dynType));
+    m_item->undoChangeProperty(Pid::DYNAMIC_TYPE, DYN_POPUP_PAGES[page][index].dynType);
+    endCommand();
+
+    if (m_item->flag(ElementFlag::IS_PREVIEW)) {
+        m_item->setFlag(ElementFlag::IS_PREVIEW, false);
+    }
+
+    INotationInteractionPtr interaction = currentNotation()->interaction();
+
+    interaction->flipHairpinsType(toDynamic(m_item));
+
+    // Hide the bounding box which appears when called using Ctrl+D shortcut
+    if (interaction->isTextEditingStarted()) {
+        interaction->endEditText();
+        interaction->startEditGrip(m_item, Grip::DRAG);
```

Comment:
Should `startEditGrip` only be called in this `if` block, or should it be called regardless of whether text editing was active?

Reply:
`startEditGrip` is already called when we select an existing dynamic, so that's why we are calling it for the empty popup only which is in text editing mode.